### ISSUE TITLE
NodeJS version does not work out of the box

### ISSUE THREAD

**Issue by mLuby** [bpampuch/pdfmake#1312]
Clearly Node ("server-side") setup is a stumbling block, especially around fonts.
#916
https://github.com/bpampuch/pdfmake/issues/1044#issuecomment-308321337
#1062 

I'd say libraries should just work out of the box. Specifically, this should work:
```js
var PdfMake = require('pdfmake')
var fs = require('fs')

const doc = new PdfMake().createPdfKitDocument({content: 'my text'})
doc.pipe(fs.createWriteStream('myFile.pdf'))
doc.end()
```
Unfortunately, this simple example errors out with `Error: Font 'Roboto' in style 'normal' is not defined in the font section of the document definition` because `PdfMake` _requires_ a font definition object. I think that's something that should be **provided by default**, and can be **optionally overwritten** by people who want custom fonts. Doing so would make using this library much simpler to use with Node. This is the minimal example that works:
```js
var PdfPrinter = require('pdfmake');
var fs = require('fs');

const doc = new PdfPrinter({
	Roboto: {normal: new Buffer(require('pdfmake/build/vfs_fonts.js').pdfMake.vfs['Roboto-Regular.ttf'], 'base64')}
}).createPdfKitDocument({content: 'my text'})
doc.pipe(fs.createWriteStream('myFile.pdf'))
doc.end()
```
Perhaps it should be in [the wiki](https://github.com/bpampuch/pdfmake/wiki/Building-font-file-via-shell-script).

**Comment by mozfet** [comment]
After some digging, I leaned that PDFKit, which pdfmake is built on, includes support for 14 standard fonts that are built into the pdf standard. Using these fonts avoid embedding any fonts in the document, thus also use less bandwidth. Since I run meteor on production clusters, I stay away from using the file system and use a buffer stream for my example instead.

Thus this makes the minimum node (es6) example I can create:
```
import PdfPrinter from 'pdfmake';
import streamBuffers from 'stream-buffers';
const doc = new PdfPrinter({timesRoman:{normal: 'Times-Roman'})
    .createPdfKitDocument({defaultStyle:{font: 'timesRoman'}}, {content: 'my text'});
const writableBuffer = new streamBuffers.WritableStreamBuffer();
doc.pipe(writableBuffer);
doc.end();
```

I do not see any errors on the server console, and on the client Adobe Acrobat does not generate any warnings for docs created this way... but neither shows any text... for me at least...

**Comment by mozfet** [comment]
I concluded perhaps generating pdf server side is not such a good idea because I cannot get it to work, so I looked for a workaround. Now I generate the document definition on the server, and return it to the client to generate the pdf with pdfmake there. It works well so far.

**Comment by mozfet** [comment]
The workaround stopped working well... well because, because when I want a footer with a page number in it, it is defined as footer key in the document definition as a function to dynamically determine the content of the footer. Functions are not very portable and tricky to pass from server to client, and in my case using meteor methods to transfer the document definition does not work, so my reports cannot have a page number.

**Comment by mozfet** [comment]
The workaround for the workaround is to create the document definition, excluding the footer on the server, and to add the footer on the client.

**Comment by FrancisCalizo** [comment]
@mozfet Hey Mozfet, do you happen to have code for me to get an idea on how to get a pdf generated server-side? I've been stuck on this for a bit and just need some sort of idea. I am using Node and Express at the moment and am confused as to how to even get a pdf generated with these docs.

**Comment by mLuby** [comment]
@FrancisCalizo does my second code block above not work for you?

@mozfet I remember seeing mention of fonts when tracing this issue through pdfmake's dependencies but I don't remember actually ever finding the font definitions. You think one or more deps actually come packaged with font files?

**Comment by honestserpent** [comment]
I can't make it work even using:

```
var pdfMake = require('pdfmake');
var docDefinition = { content: 'This is a test' };
const pdfDocGenerator = pdfMake.createPdf(docDefinition);
```
I get a `TypeError: pdfMake.createPdf is not a function`
Does anyone have an idea about what am I messing up?

**Comment by honestserpent** [comment]
@mLuby for me it works, but I don't need to create a file. I need to create a Buffer in order to be able to send the generated pdf as an email attachment. Do you have any idea on how to do it?

**Comment by FrancisCalizo** [comment]
@mLuby thanks for the reply. I am sort somewhat unfamiliar with some backend aspects, I've been studying JS for about 3-4 months. 

I created a file using Express Generator. Would that second block of code go into a file in the routes folder? I used the second block of code in the index.js file in the routes folder. 

Sorry for the newbie question, still trying to get a hang of all this.

**Comment by honestserpent** [comment]
@FrancisCalizo it doesn't matter where you put it. You could put it in a route handler in an Express application, in a standalone script or whatever else.

For instance, if you want to generate and save a PDF file when you call a certain API, you could do something like this:

```
[create express server]
var PdfPrinter = require('pdfmake');
var fs = require('fs');

app.get('/generatePDF', (req, res)=>{
  const doc = new PdfMake().createPdfKitDocument({content: 'my text'})
  doc.pipe(fs.createWriteStream('myFile.pdf'))
  doc.end()
})
```

**Comment by mozfet** [comment]
@mLuby I never got it working server side... I generate the document definition server side and the pdf client side...

**Comment by mozfet** [comment]
@honestserpent I would not expect that code to work server side because PdfMake is never imported, PdfPrinter is. I could not find a PdfMake function available for import server side (I logged the imported object to the console to check), and got an error when trying to use pdf printer without a fonts object.



As I understand it PdfMake is for use client side, PdfPrinter is used server side.

**Comment by mozfet** [comment]
@mLuby I think that you will not find these standard fonts in the pdfmake dependency chain because they are embedded in the pdf viewer binaries already installed as client-side software, they are thus built in and need not be embedded, are thus only referred to by name.

**Comment by honestserpent** [comment]
it doens't work because I messed it up. [Here is a working gist](https://gist.github.com/honestserpent/657d9e471e0cf54ea583d419375c2d66)

**Comment by honestserpent** [comment]
@mozfet @mLuby I also made a first attempt to create the express version. [Here is the gist](https://gist.github.com/honestserpent/1f223fd8e5ec55ba51a4cb81c4c0a602).

**Comment by mozfet** [comment]
@honestserpent I got that far as well, which is basically your opening post. Then the pdf downloads on the client side, it complains about missing font. When I used the standard fonts, then there was no error in pdf viewer, but nothing was visible either. The meta information is however clearly readable inside adobe as file properties, and thus I do not expect it to be a transmission encoding issue...

**Comment by honestserpent** [comment]
@mozfet the one I did is purely server side, and it does not give me any problems with the generated PDFs. Gotta say that the generated pdf is VERY simple (just a line of text). But hey, it works! :D

**Comment by mozfet** [comment]
@honestserpent I have created a reply [gist](https://gist.github.com/mozfet/1fcaf22e9351a08b33e1dcf88ff2126d) of doing the same on Meteor. To me it looks like I am doing the same as you...

**Comment by mozfet** [comment]
Nooo... <HANDS IN HAIR>


When I download using Chrome and open the generated pdf on OSX using Adobe Acrobat I get an error.
![screen shot 2018-04-19 at 17 21 59](https://user-images.githubusercontent.com/5113240/39001477-ab54fe8c-43f6-11e8-88ac-3f1e0cde2160.png)

**Comment by honestserpent** [comment]
Well, I don't know specifically. Have you tried my example? Try to implement it in meteor and let me know. I can try meteor but not now

**Comment by mozfet** [comment]
@honestserpent It is all based on your example, just refactored. Anyhoo I do not need it now anymore, it is working ok with the workaround, and I'm starting to think it may even be a good thing doing it this way, less server processing means faster responses and it is quite fast enough on the client. Perhaps in a few weeks I give it another go.

**Comment by honestserpent** [comment]
I am wondering if this is an issue with Mac OS. Do you have a chance to try it on a different OS?

**Comment by mozfet** [comment]
I do not have another OS instance read ATM... here is the file; can you open it and see any text?
[report.pdf](https://github.com/bpampuch/pdfmake/files/1931074/report.pdf)

**Comment by honestserpent** [comment]
No, I can open it but I don't see any text. But it makes sense. Would be interesting to test your code on a different platform. I'm gonna test it on mine.

**Comment by meta7x** [comment]
Thanks for the hint! This was the final piece in the puzzle of generating a PDF on the server. It was quite confusing that I had to load the fonts on my own when they were just specified as a path in every example as well as in the documentation.

**Comment by mLuby** [comment]
@bpampuch would you welcome a PR to fix this? I believe it's a non-breaking API change.

**Comment by liborm85** [comment]
@mLuby, I do not see any bug to fix, because: 
* Wiki page [Building font file via shell script](https://github.com/bpampuch/pdfmake/wiki/Building-font-file-via-shell-script) describe building custom fonts for client-side. Not for server-side.
* In server-side do not use virtual file system in `vfs_fonts.js`, use real font files.
* How to use pdfmake on server-side and how to set fonts see: [examples/basics.js](https://github.com/bpampuch/pdfmake/blob/master/examples/basics.js)
* Fonts to use on server side is available in [examples/fonts](https://github.com/bpampuch/pdfmake/tree/master/examples/fonts) (any other font files may be used)

**Comment by mLuby** [comment]
It's not a bug exactly, just a frustrating user experience (and others agree 👍).

- The wiki says "pdfMake uses the Roboto font by default" which is incorrect; pdfMake uses no fonts by default (server-side at least). This makes it confusing when the user gets an error saying pdfMake can't find Roboto.
- it's highly irregular for the user to have to do a special build step to get an npm-installed library to work, especially a step [using bash](
https://github.com/bpampuch/pdfmake/wiki/Building-font-file-via-shell-script).
- The fact that it's not obvious you have to do this build step from the [README](https://github.com/bpampuch/pdfmake#getting-started) adds to the confusion.
- [examples/basics.js](
https://github.com/bpampuch/pdfmake/blob/master/examples/basics.js) doesn't work because [examples/fonts](https://github.com/bpampuch/pdfmake/tree/master/examples/fonts/) are stripped away by [.npmignore](https://github.com/bpampuch/pdfmake/blob/master/.npmignore#L5)

My goal here is to save others the time and frustration I went through to get this library working. Once it works, pdfMake is a _super_ helpful piece of software, so **thank you**! 😄

It's your work, so I'll close this issue if you still disagree with me that the UX could be improved.

**Comment by muffeeee** [comment]
This module is a nightmare for server-side. You should really just remove it from npm so people don't have to waste their time on this until you can provide proper (and functional) documentation. It doesn't even seem like you really understand your own project..

Even after two hours of digging in GitHub issues and a couple stackoverflow questions, the library still wouldn't render a PDF document with my specified font. Tried doing it the way the docs said (if you can even call that docs, half the functionality of this project aren't even mentioned), tried converting the fonts to base64 using the vfs_fonts.js method (as described in the comments here).

I'll just find something else to generate my PDFs, I guess.

**Comment by meta7x** [comment]
You’re right, the features of this module are great, but getting it to work server side is a hassle and the documentation doesn’t help whatsoever, which is a shame..

**Comment by mLuby** [comment]
@muffeeee It's still a pretty darn useful piece of software, but I 💯agree with you that the documentation and dev experience could be significantly improved for server-side users.

Makes me wonder whether there's support for a server-focused fork? 👍👎

**Comment by kaushikpathak** [comment]
> This module is a nightmare for server-side. You should really just remove it from npm so people don't have to waste their time on this until you can provide proper (and functional) documentation. It doesn't even seem like you really understand your own project..
> 
> Even after two hours of digging in GitHub issues and a couple stackoverflow questions, the library still wouldn't render a PDF document with my specified font. Tried doing it the way the docs said (if you can even call that docs, half the functionality of this project aren't even mentioned), tried converting the fonts to base64 using the vfs_fonts.js method (as described in the comments here).
> 
> I'll just find something else to generate my PDFs, I guess.

Hey, @muffeeee 
I am using pdfmake for server side and I have used multiple fonts, even I am using MICR for printing cheque. Just a few changes you need to do in code of pdfmake. 
I agree pdfmake is not that good but we can modify code as per our use.

**Comment by YoannBuzenet** [comment]
Hey ! So could anyone make a small recap on how to make it work serverside ? I'm stuck on the font problem and I still don't get how to proceed.

Thank you :)

**Comment by YoannBuzenet** [comment]
Just to share with you guys What I did : 

- I took the Roboto package from NPM and installed it. ("roboto-font": "0.1.0")
- Wrote the loading command `(let printer = new pdfMake(fonts))` **AFTER** the font object declaration

And I finally stopped having this message...

Credits to this SO answer : 
https://stackoverflow.com/questions/45196528/how-to-create-a-pdf-on-node-js-using-pdfmake-and-vfs-fonts

`let fonts = {
    Roboto: {
        normal: 'node_modules/roboto-font/fonts/Roboto/roboto-regular-webfont.ttf',
        bold: 'node_modules/roboto-font/fonts/Roboto/roboto-bold-webfont.ttf',
        italics: 'node_modules/roboto-font/fonts/Roboto/roboto-italic-webfont.ttf',
        bolditalics: 'node_modules/roboto-font/fonts/Roboto/roboto-bolditalic-webfont.ttf'
    }
};
let printer = new pdfMake(fonts);`

**Comment by paddotk** [comment]
[This](https://github.com/bpampuch/pdfmake/issues/1894#issuecomment-573292051) answer seems to work. However I think pdfMake should be doing this by default, I almost gave up trying..

**Comment by govnag** [comment]
@mLuby Thank you!! Your buffer example is the only thing that worked for me; including adding the @YoannBuzenet fix for including the roboto font

I updated `new Buffer()` -> `Buffer.from()` and it still works

**Comment by liborm85** [comment]
In version 0.3.0 is available Roboto fonts in npm release in folder `fonts`. How to use pdfmake and fonts on server is described in the documentation and examples.
